[clang-format] Add SpacesInParensOption for filtering repeated parens

gedare opened the pull request:
The __attribute((specifier-list)) currently is formatted based on the SpacesInParensOptions.Other (previously, SpacesInParentheses). This change allows finer control over addition of spaces between the consecutive parens, and between the inner parens and the list of attribute specifiers.

Differential Revision: https://reviews.llvm.org/D155529

This is migrated from Phabricator, see more discussion there.

gedare committed:
Add SpaceInParensOption ExceptDoubleParentheses

This change allows finer control over addition of spaces between a pair of
consecutive opening parentheses with a pair of closing parentheses.

gedare commented:
> This is migrated from Phabricator, see more discussion there. I will next provide an option for SpacesInParensOptions.Doubled to control (and override) injection of spaces within doubled parens `(( ))`.

Actually, spaces in double parens might not work out well. It will not allow separation of multiple parens in some contexts such as conditional blocks or function calls. It may be better to provide a sub-option to this `InAttributeSpecifiers` instead.

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -16815,6 +16817,18 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
+      FormatStyle::SIPCS_Always;
+  verifyFormat("SomeType *__attribute__( ( attr ) ) *a = NULL;", Spaces);
+  verifyFormat("void __attribute__( ( naked ) ) foo(int bar)", Spaces);
+  verifyFormat("void f( ) __attribute__( ( asdf ) );", Spaces);
+  Spaces.SpacesInParensOptions.InAttributeSpecifiers =
```
```suggestion
  verifyFormat("void f( ) __attribute__( ( asdf ) );", Spaces);
  
  Spaces.SpacesInParensOptions.InAttributeSpecifiers =
```
For a better readability.

owenca commented:
> The __attribute((specifier-list)) currently is formatted based on the SpacesInParensOptions.Other (previously, SpacesInParentheses). This change allows finer control over addition of spaces between the consecutive parens, and between the inner parens and the list of attribute specifiers.
> 
> Differential Revision: https://reviews.llvm.org/D155529
> 
> This is migrated from Phabricator, see more discussion there.

I expressed my opinion there:

> I would have no problem if this new option is extended to handle all double parens, e.g. if (( i = j )), decltype(( x )), etc.

So I still prefer that we have a boolean suboption (e.g. `ConsecutiveParentheses`) that covers all double parens.

mydeveloperday reviewed (approved):
LGTM

HazardyKnusperkeks reviewed (approved):
But Owen's proposal isn't bad either.

gedare commented:
> > The __attribute((specifier-list)) currently is formatted based on the SpacesInParensOptions.Other (previously, SpacesInParentheses). This change allows finer control over addition of spaces between the consecutive parens, and between the inner parens and the list of attribute specifiers.
> > Differential Revision: https://reviews.llvm.org/D155529
> > This is migrated from Phabricator, see more discussion there.
> 
> I expressed my opinion there:
> 
> > I would have no problem if this new option is extended to handle all double parens, e.g. if (( i = j )), decltype(( x )), etc.
> 
> So I still prefer that we have a boolean suboption (e.g. `ConsecutiveParentheses`) that covers all double parens.

I will provide something shortly along these lines by duplicating the logic I used for this `InAttributeSpecifiers` to apply to the other suboptions of `SpacesInParensOptions`. Covering all double parens in a single sub-option is not precise enough for code bases that may want to have `__attribute__(( x ))` but also allow `if ( ( x ) )`.

gedare commented:
I'm still making changes on this. Not sure how to mark that in GH. Need to add test cases for double parens, and get it working for the newly added suboptions.

gedare commented:
> > The __attribute((specifier-list)) currently is formatted based on the SpacesInParensOptions.Other (previously, SpacesInParentheses). This change allows finer control over addition of spaces between the consecutive parens, and between the inner parens and the list of attribute specifiers.
> > Differential Revision: https://reviews.llvm.org/D155529
> > This is migrated from Phabricator, see more discussion there.
> 
> I expressed my opinion there:
> 
> > I would have no problem if this new option is extended to handle all double parens, e.g. if (( i = j )), decltype(( x )), etc.
> 
> So I still prefer that we have a boolean suboption (e.g. `ConsecutiveParentheses`) that covers all double parens.

Please have a look, I have moved this PR more toward this direction providing fine-grained controls over double parens.

Someone will have to add the decltype understanding in order to make `decltype(( x ))` work without having to enable `Other.NonConsecutive`.

HazardyKnusperkeks commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -4007,26 +3992,45 @@ bool TokenAnnotator::spaceRequiredBetween(const AnnotatedLine &Line,
+    const auto AddSpace = [&](FormatStyle::SpacesInParensCustomStyle Option) {
+      if (Option == FormatStyle::SIPCS_Always)
+        return true;
+      if (Option == FormatStyle::SIPCS_NonConsecutive) {
+        if (NotConsecutiveParens(LeftParen, RightParen))
+          return true;
+      }
```
```suggestion
      if (Option == FormatStyle::SIPCS_NonConsecutive
        && NotConsecutiveParens(LeftParen, RightParen)) {
          return true;
      }
```

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -17016,6 +17056,158 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
+  verifyFormat("decltype( ( x ) ) y = z;", Spaces);
+  verifyFormat("decltype( ( foo() ) ) a = foo();", Spaces);
+  verifyFormat("decltype( ( bar( 10 ) ) ) a = bar( 11 );", Spaces);
+  verifyFormat("x = foo( ( a * ( b - c ) ) );", Spaces);
+  Spaces.SpacesInParensOptions.Other = FormatStyle::SIPCS_NonConsecutive;
```
```suggestion
  verifyFormat("x = foo( ( a * ( b - c ) ) );", Spaces);
  
  Spaces.SpacesInParensOptions.Other = FormatStyle::SIPCS_NonConsecutive;
```

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -17016,6 +17056,158 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
+  verifyFormat("decltype(( x )) y = z;", Spaces);
+  verifyFormat("decltype((foo())) a = foo();", Spaces);
+  verifyFormat("decltype((bar( 10 ))) a = bar( 11 );", Spaces);
+  verifyFormat("x = foo((a * ( b - c )));", Spaces);
+  Spaces.SpacesInParensOptions.Other = FormatStyle::SIPCS_Never;
```
```suggestion
  verifyFormat("x = foo((a * ( b - c )));", Spaces);
  
  Spaces.SpacesInParensOptions.Other = FormatStyle::SIPCS_Never;
```

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -17001,6 +17056,158 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
+               "default:\n"
+               "  break;\n"
+               "}",
+               Spaces);
+  Spaces.SpacesInParensOptions.InConditionalStatements =
```
```suggestion
               Spaces);
               
  Spaces.SpacesInParensOptions.InConditionalStatements =
```

owenca commented:
> Covering all double parens in a single sub-option is not precise enough for code bases that may want to have `__attribute__(( x ))` but also allow `if ( ( x ) )`.

We should not go overboard with supporting all kinds of options/suboptions imaginable. I don't think we should support spaces between consecutive parens in the first place, which was likely a bug.

We can't just disregard the long-standing [policy](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#adding-additional-style-options) on adding new options. If anyone thinks that policy is outdated, they should go through the RFC process and get it updated.

gedare commented:
> > Covering all double parens in a single sub-option is not precise enough for code bases that may want to have `__attribute__(( x ))` but also allow `if ( ( x ) )`.
> 
> We should not go overboard with supporting all kinds of options/suboptions imaginable. I don't think we should support spaces between consecutive parens in the first place, which was likely a bug.
> 
Well, it seems to be far too late to go back and change the default behavior.

> We can't just disregard the long-standing [policy](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#adding-additional-style-options) on adding new options. If anyone thinks that policy is outdated, they should go through the RFC process and get it updated.

I have no problem with this policy. I think it is a good policy. I can advocate for my position. My patches relate to the style maintenance of a widely-used real-time operating system. The code is [self-hosted](https://git.rtems.org/rtems/) and [mirrored on GitHub](https://github.com/RTEMS/rtems) with a [documented style guide](https://docs.rtems.org/branches/master/eng/coding-formatting.html). In addition, I have been and will continue to be willing to provide bug fix and other maintenance support to clang-format based on the changes submitted to support these style needs.

At the moment, I need to satisfy a particular style rule that requires me to add spaces inside of conditional and compound expressions. This support basically exists in `clang-format` but the compound expressions fall under the `SpacesInParensOption.Other` category. In order to control compound expressions separately from other uses of parens, it seems to be the easiest thing to explicitly define "everything else" and control the injection of spaces within them. Eventually, it results in a much more controlled and configurable way to manage whitespace.

If this is not desired, I will need to step back to consider my options further, and probably have to take this up with the other maintainers of my code base to see what, if any, concessions we can make based on what is allowed with `clang-format`. I'm trying to get to a point where we are able to use `clang-format` on our code base with enforced CI actions. It is still a ways to go before I get there.

owenca commented:
> The code is [self-hosted](https://git.rtems.org/rtems/) and [mirrored on GitHub](https://github.com/RTEMS/rtems) with a [documented style guide](https://docs.rtems.org/branches/master/eng/coding-formatting.html). In addition, I have been and will continue to be willing to provide bug fix and other maintenance support to clang-format based on the changes submitted to support these style needs.
> 
> At the moment, I need to satisfy a particular style rule that requires me to add spaces inside of conditional and compound expressions.

It seems the RTEMS style guide you linked above doesn't have any examples for conditional expressions and compound expressions, and by _conditional expressions_ I assume it means conditionals of control statements rather than ternary [conditional expressions](https://www.ibm.com/docs/en/zos/3.1.0?topic=operators-conditional-expressions).

Would `__attribute__((noreturn))`, `if ((i = j))`, `decltype((x))`, and `while (((i + 1) * j - 2) * k > 3)` be formatted as `__attribute__(( noreturn ))`, `if (( i = j ))`, `decltype(( x ))`, and `while ( ( ( i + 1 ) * j - 2 ) * k > 3 )`, respectively?

gedare commented:
> > The code is [self-hosted](https://git.rtems.org/rtems/) and [mirrored on GitHub](https://github.com/RTEMS/rtems) with a [documented style guide](https://docs.rtems.org/branches/master/eng/coding-formatting.html). In addition, I have been and will continue to be willing to provide bug fix and other maintenance support to clang-format based on the changes submitted to support these style needs.
> > At the moment, I need to satisfy a particular style rule that requires me to add spaces inside of conditional and compound expressions.
> 
> It seems the RTEMS style guide you linked above doesn't have any examples for conditional expressions and compound expressions, and by _conditional expressions_ I assume it means conditionals of control statements rather than ternary [conditional expressions](https://www.ibm.com/docs/en/zos/3.1.0?topic=operators-conditional-expressions).
> 
Correct. Unfortunately, no one (myself) has not taken the time to thoroughly document all examples. The existing style has been inherited from the original author of our code base, as these things tend to be for older projects.

> Would `__attribute__((noreturn))`, `if ((i = j))`, `decltype((x))`, and `while (((i + 1) * j - 2) * k > 3)` be formatted as `__attribute__(( noreturn ))`, `if (( i = j ))`, `decltype(( x ))`, and `while ( ( ( i + 1 ) * j - 2 ) * k > 3 )`, respectively?

Almost, it would also be `if ( ( i = j ) )`, Although repeated parens are typically removed. So a better example is `if ( ( i = j ) && ( k > 3 ) )`

We don't have `decltype` in our C code base. I can add support for it as an option if desired.

We also have cases such as `__attribute__(( __aligned__( _alignment ) ))`

owenca commented:
> > Would `__attribute__((noreturn))`, `if ((i = j))`, `decltype((x))`, and `while (((i + 1) * j - 2) * k > 3)` be formatted as `__attribute__(( noreturn ))`, `if (( i = j ))`, `decltype(( x ))`, and `while ( ( ( i + 1 ) * j - 2 ) * k > 3 )`, respectively?
> 
> Almost, it would also be `if ( ( i = j ) )`, Although repeated parens are typically removed. So a better example is `if ( ( i = j ) && ( k > 3 ) )`
> 
> We don't have `decltype` in our C code base. I can add support for it as an option if desired.
> 
> We also have cases such as `__attribute__(( __aligned__( x ) ))`

I'm ok with all of the above except for `if ( ( i = j ) )`, which should be formatted as `if (( i = j ))` because the extra pair of parentheses is not superfluous. This is similar to `decltype(( x))`, and the`RemoveParentheses` option doesn't strip the extra pair for either.

It seems adding a boolean sub-option that targets double pairs of parentheses as I suggested before is feasible although a better name than `ConsecutiveParentheses` may be needed.

gedare commented:
> > > Would `__attribute__((noreturn))`, `if ((i = j))`, `decltype((x))`, and `while (((i + 1) * j - 2) * k > 3)` be formatted as `__attribute__(( noreturn ))`, `if (( i = j ))`, `decltype(( x ))`, and `while ( ( ( i + 1 ) * j - 2 ) * k > 3 )`, respectively?
> > 
> > 
> > Almost, it would also be `if ( ( i = j ) )`, Although repeated parens are typically removed. So a better example is `if ( ( i = j ) && ( k > 3 ) )`
> > We don't have `decltype` in our C code base. I can add support for it as an option if desired.
> > We also have cases such as `__attribute__(( __aligned__( x ) ))`
> 
> I'm ok with all of the above except for `if ( ( i = j ) )`, which should be formatted as `if (( i = j ))` because the extra pair of parentheses is not superfluous. This is similar to `decltype(( x ))`, and the`RemoveParentheses` option doesn't strip the extra pair for either.
> 
Yes, good point.

> It seems adding a boolean sub-option that targets double pairs of parentheses as I suggested before is feasible although a better name than `ConsecutiveParentheses` may be needed.

I'm fine to fix but request a concrete suggestion or proposal before I spend much more time on this. It might also work to suppress optionally the additional space on the first/last parenthesis pair, for these special "paired" parens. I think the way I already worked this patch is the right way to go to preserve backward compatibility while still allowing control over the repeated parentheses. 

Looks like it is called `MultipleParentheses` in the `RemoveParentheses` option. I should rename to that.

owenca commented:
> > It seems adding a boolean sub-option that targets double pairs of parentheses as I suggested before is feasible although a better name than `ConsecutiveParentheses` may be needed.
> 
> I'm fine to fix but request a concrete suggestion or proposal before I spend much more time on this. It might also work to suppress optionally the additional space on the first/last parenthesis pair, for these special "paired" parens. I think the way I already worked this patch is the right way to go to preserve backward compatibility while still allowing control over the repeated parentheses.
> 
> Looks like it is called `MultipleParentheses` in the `RemoveParentheses` option. I should rename to that.

My idea was:
- Keep the existing boolean suboptions and add a new one. Let's call it `ExceptDoubleParentheses` for now.
- If it's set to true, assuming all other suboptions are set to true, the above examples would be formatted as:
```
__attribute__(( noreturn ))
__attribute__(( __aligned__( x ) ))

if (( i = j ))
if ( ( i = j ) && ( k > 3 ) ) // not affected by ExceptDoubleParentheses

decltype(( x ))

while ( ( ( i + 1 ) * j - 2 ) * k > 3 ) // not affected by ExceptDoubleParentheses
```
- If it's set to false, assuming all other suboptions are set to true, the above examples would be formatted as:
```
__attribute__( ( noreturn ) )
__attribute__( ( __aligned__( x ) ) )

if ( ( i = j ) )
if ( ( i = j ) && ( k > 3 ) ) // not affected by ExceptDoubleParentheses

decltype( ( x ) )

while ( ( ( i + 1 ) * j - 2 ) * k > 3 ) // not affected by ExceptDoubleParentheses
```

This way, we would avoid making all suboptions an `enum` and repeating `NonConsecutive` in every suboption.

gedare commented:
> > > It seems adding a boolean sub-option that targets double pairs of parentheses as I suggested before is feasible although a better name than `ConsecutiveParentheses` may be needed.
> > 
> > 
> > I'm fine to fix but request a concrete suggestion or proposal before I spend much more time on this. It might also work to suppress optionally the additional space on the first/last parenthesis pair, for these special "paired" parens. I think the way I already worked this patch is the right way to go to preserve backward compatibility while still allowing control over the repeated parentheses.
> > Looks like it is called `MultipleParentheses` in the `RemoveParentheses` option. I should rename to that.
> 
> My idea was:
> 
> * Keep the existing boolean suboptions and add a new one. Let's call it `ExceptDoubleParentheses` for now.
> * If it's set to true, assuming all other suboptions are set to true, the above examples would be formatted as:
> 
> ```
> __attribute__(( noreturn ))
> __attribute__(( __aligned__( x ) ))
> 
> if (( i = j ))
> if ( ( i = j ) && ( k > 3 ) ) // not affected by ExceptDoubleParentheses
> 
> decltype(( x ))
> 
> while ( ( ( i + 1 ) * j - 2 ) * k > 3 ) // not affected by ExceptDoubleParentheses
> ```
> 
> * If it's set to false, assuming all other suboptions are set to true, the above examples would be formatted as:
> 
> ```
> __attribute__( ( noreturn ) )
> __attribute__( ( __aligned__( x ) ) )
> 
> if ( ( i = j ) )
> if ( ( i = j ) && ( k > 3 ) ) // not affected by ExceptDoubleParentheses
> 
> decltype( ( x ) )
> 
> while ( ( ( i + 1 ) * j - 2 ) * k > 3 ) // not affected by ExceptDoubleParentheses
> ```
> 
> This way, we would avoid making all suboptions an `enum` and repeating `NonConsecutive` in every suboption.

OK, I think I understand. This would be greatly simplified if we can annotate the `DoubleParenthesis` situation inside of `UnwrappedLineParser::parseParens`.

gedare commented:
* Rebased to resolve conflict in ReleaseNotes.
* Added `ExceptDoubleParentheses` as proposed by @owenca 

This seems to work reasonably well for the cases I care about, as far as I can tell.

HazardyKnusperkeks commented on clang/include/clang/Format/Format.h:
```diff
@@ -4542,18 +4542,51 @@ struct FormatStyle {
-    ///    true:                                  false:
-    ///    if ( a )  { ... }              vs.     if (a) { ... }
-    ///    while ( i < 5 )  { ... }               while (i < 5) { ... }
+    ///    true:
```
I like the previous version way better.

HazardyKnusperkeks commented on clang/include/clang/Format/Format.h:
```diff
@@ -4542,18 +4542,51 @@ struct FormatStyle {
   ///     Other: true
```
Add the options here too?

HazardyKnusperkeks commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -4006,11 +3993,58 @@ bool TokenAnnotator::spaceRequiredBetween(const AnnotatedLine &Line,
+                     (HasEqualBeforeNextParen() &&
+                      (LPrev->isOneOf(tok::kw_if, tok::kw_while) ||
+                       LPrev->endsSequence(tok::kw_constexpr, tok::kw_if)))));
+      return !(DoubleParens && SuppressSpace);
```
```suggestion
      return !DoubleParens || !SuppressSpace;
```

HazardyKnusperkeks commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -4006,11 +3993,58 @@ bool TokenAnnotator::spaceRequiredBetween(const AnnotatedLine &Line,
+      const auto *RPrev = RightParen ? RightParen->Previous : nullptr;
+      const auto *LNext = LeftParen->Next;
+      const auto *LPrev = LeftParen->Previous;
+      const bool DoubleParens =
+          RPrev && RPrev->is(tok::r_paren) && LNext && LNext->is(tok::l_paren);
+      auto HasEqualBeforeNextParen = [&]() {
```
```suggestion
      if (RPrev && RPrev->is(tok::r_paren) && LNext && LNext->is(tok::l_paren))
        return false;
      auto HasEqualBeforeNextParen = [&]() {
```
Don't calculate all the stuff, if not needed.

HazardyKnusperkeks commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -4006,11 +3993,58 @@ bool TokenAnnotator::spaceRequiredBetween(const AnnotatedLine &Line,
+    const auto IsAttributeParen = [](const FormatToken *Paren) {
+      return Paren && Paren->isOneOf(TT_AttributeLParen, TT_AttributeRParen);
+    };
+    auto AddSpaceInDoubleParens = [&]() {
```
Isn't it more like AddSpace**Except**InDoubleParens?

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -4542,18 +4542,51 @@ struct FormatStyle {
+    ///    false:
+    ///    _attribute__((noreturn))
+    /// \endcode
+    bool InAttributeSpecifiers;
```
I still think this is too specific. It should be included in `Other` below.

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -4569,23 +4602,40 @@ struct FormatStyle {
-    ///    t f( Deleted & ) & = delete;   vs.     t f(Deleted &) & = delete;
+    ///   true:
+    ///   t f( Deleted & ) & = delete;
+    ///   decltype( ( x ) )
```
```suggestion
    ///    __attribute__( ( noreturn ) )
    ///   decltype( ( x ) )
```

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -4569,23 +4602,40 @@ struct FormatStyle {
+    /// \code
+    ///   false:
+    ///   t f(Deleted &) & = delete;
+    ///   decltype((x))
```
```suggestion
    ///    __attribute__((noreturn))
    ///   decltype((x))
```

a1batross commented:
Any news on this?

a1batross commented:
I have built clang-format with these changes. On current commit, it sadly doesn't remove the space in these patterns:
```c
memset( g_used, 0, sizeof( g_used ) );
if( FBitSet( ent->curstate.effects, EF_INVLIGHT ) )
```

After reverting last commit, it got somewhat closer:
```c
memset(g_used, 0, sizeof( g_used ));
if(FBitSet( ent->curstate.effects, EF_INVLIGHT ))
```

But is there a way to configure it like this?
```c
memset( g_used, 0, sizeof( g_used ))
if( FBitSet( ent->curstate.effects, EF_INVLIGHT ))
```

gedare commented:
I saw some comments on this. Looks like some were deleted maybe. This is still in my queue, but unclear when I'll get back to it.

HazardyKnusperkeks commented on clang/include/clang/Format/Format.h:
```diff
@@ -4644,22 +4644,36 @@ struct FormatStyle {
     /// \code
     ///    true:                                  false:
-    ///    if ( a )  { ... }              vs.     if (a) { ... }
+    ///    if ( ( a ) )  { ... }          vs.     if ((a)) { ... }
```
Don't change this.

If you want to show your option, adapt its documentation.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -17265,6 +17317,268 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
+  verifyFormat("decltype(( foo() )) a = foo();", Spaces);
+  verifyFormat("decltype(( bar( 10 ) )) a = bar( 11 );", Spaces);
+  verifyFormat("decltype(( foo->bar )) baz;", Spaces);
+  verifyFormat("if (( i = j ))\n"
+               "  do_something( i );",
+               Spaces);
+  verifyFormat("if constexpr (( a = b ))\n"
+               "  c;",
+               Spaces);
```
There should be a space between the double parens because `ExceptDoubleParentheses` is false.

owenca reviewed (commented):
I think we can simply check if  a sequence of two `l_parens` have a matching sequence of two `r_parens` and insert spaces only if `ExceptDoubleParentheses` is false. See #93439.

owenca committed:
Improve the logic.

gedare committed:
FormatTest: fix and add test cases for ExceptDoubleParentheses

gedare commented:
I had to rebase for merge conflicts, so I also cleaned up the commit history a bit. This seems to be working fine and tests pass.

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -4650,10 +4650,23 @@ struct FormatStyle {
   ///     Other: true
   /// \endcode
   struct SpacesInParensCustom {
+    /// Override any of the following options to prevent addition of space
+    /// between the first two parentheses in situations where a pair of
+    /// parentheses have been used.
```
This seems unclear or imprecise.

gedare replied:
Reworded

owenca reviewed (commented):
Do we need all these new test cases? If yes, consider moving the test (and other `ConfigurableSpacesIn...` tests) to a separate file.

a1batross commented:
> I saw some comments on this. Looks like some were deleted maybe. This is still in my queue, but unclear when I'll get back to it.

Yeah, sorry, I posted a comment here but decided it was a bit irrelevant. :)

But going to ask anyway then. Can this setting remove the space from repeated parentheses pattern like this?

```c
foo( (void *)buf, size );
```

to:

```c
foo((void *)buf, size );
```

So it does remove it on one side where the C-style cast parentheses, but doesn't on the other side?

gedare committed:
Address review comments

gedare commented:
> Do we need all these new test cases? If yes, consider moving the test (and other `ConfigurableSpacesIn...` tests) to a separate file.

I have simplified the additional test cases.

gedare commented:
> > I saw some comments on this. Looks like some were deleted maybe. This is still in my queue, but unclear when I'll get back to it.
> 
> Yeah, sorry, I posted a comment here but decided it was a bit irrelevant. :)
> 
> But going to ask anyway then. Can this setting remove the space from repeated parentheses pattern like this?
> 
> ```c
> foo( (void *)buf, size );
> ```
> 
> to:
> 
> ```c
> foo((void *)buf, size );
> ```
> 
> So it does remove it on one side where the C-style cast parentheses, but doesn't on the other side?

No, at present this would not be supported. You might be able to mimic the logic here to do it, but it would also need some good justification that there is an existing style that uses it.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -17125,6 +17125,23 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
+  verifyFormat("decltype((foo())) a = foo();", Spaces);
+  verifyFormat("decltype((bar(10))) a = bar(11);", Spaces);
+  verifyFormat("if ((x - y) && (a ^ b))\n"
+               "  f();\n",
```
No newline at the end of test cases.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -17159,6 +17176,23 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
+  verifyFormat("decltype( ( foo() ) ) a = foo();", Spaces);
+  verifyFormat("decltype( ( bar( 10 ) ) ) a = bar( 11 );", Spaces);
+  verifyFormat("if ( ( x - y ) && ( a ^ b ) )\n"
+               "  f();\n",
```
Ditto.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -17203,6 +17238,23 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
+  verifyFormat("decltype((foo( ))) a = foo( );", Spaces);
+  verifyFormat("decltype((bar(10))) a = bar(11);", Spaces);
+  verifyFormat("if ((x - y) && (a ^ b))\n"
+               "  f( );\n",
```
Ditto.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -17310,6 +17362,63 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
+               "} while ((i = i + 1));",
+               Spaces);
+  verifyFormat("if ( (x - y) && (a ^ b) )\n"
+               "  f();\n",
```
Ditto.

gedare committed:
FormatTest: remove newlines at end of test cases

owenca committed:
Merge branch 'main' into spacesinparens-attr-rebase-4
